feat(app-delivery): continuous delivery for CDK apps

eladb opened the pull request:
Rewrite the app delivery library to allow full deployment of CDK apps. A CDK pipeline can be created using a command line tool called `cdk-pipeline` (which we will eventually integrate into the toolkit). This pipeline is also called "a bootstrapping pipeline". It's behavior is fixed - it builds and deploys a CDK app from a source control repo (currently only GitHub is supported).

In order to customize the deployment behavior, users can configure the bootstrapping pipeline to only deploy a deployment stack instead, which will in turn, deploy the rest of the CDK app. This approach provides full flexibility to CDK users as it offers CI/CD for both the pipeline itself and the rest of the app, with full customizability.

Fixed #308 
Fixes #1917 
Fixes #1151 

Moral support: @netanir

Co-authored-by: rix0rrr <huijbers@amazon.com>
Co-authored-by: RomainMuller <rmuller@amazon.com>
Co-authored-by: skinny85 <adamruka@amazon.com>

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

rix0rrr commented on packages/@aws-cdk/app-delivery/README.md:
```diff
@@ -1,151 +1,132 @@
-## Continuous Integration / Continuous Delivery for CDK Applications
-This library includes a *CodePipeline* composite Action for deploying AWS CDK Applications.
+# App Delivery
```
Have we made the previous method of building a CDK pipeline impossible now, or is that code still available? We're not really explaining how it works anymore at least.

I'm concerned because existing customers could be depending on this right now, and we'd break them with no real replacement. I would like this library to both export the CDK primitives to build this pipeline, as well as a the helper tool that builds a standard pipeline using those primitives from a config file, and explain both methods in the README.

The refactor is so big I can't really tell whether that is what's going on, but if not that's what I want :).

eladb replied:
This is basically a full replacement of the existing module. Start with the README file and judge if it offers the right building blocks. I’ll also see if we can expose more of the building blocks.

I’ll make sure to include a breaking change message in the commit. I don’t believe the current offering was of much value to anyone, especially given it did not support assets.

rix0rrr commented on packages/@aws-cdk/app-delivery/bin/cdk-pipeline:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
Can't really do this because it won't work on Windows.

Has to be a JavaScript file.

(And it's not "just a build script", it's in the hot path of every CDK user)

eladb replied:
Good point about Windows. I’ll convert to JavaScript.

eladb commented:
@skinny85, I've extracted the various pipeline actions from the app-delivery module into the codepipeline-actions module (`CfnSourceAction`, `CfnBuildAction` and `CfnDeployAction`).

You will see that I've implemented them as constructs, which made it much easier and less convoluted. I am not super happy that those don't look & feel like the other pipeline actions, so we should re-visit once we refactor the Action API.

Let's align the terminology:
- "Boot pipeline": the pipeline with the fixed structure that's responsible for building the app and deploying all the stacks not marked with `autoDeploy: false`.
- "App pipeline": optionally, the pipeline that apps can define)

The app-delivery module now includes only the boot pipeline (and app) and a command line program that reads a yaml file and creates/updates your boot pipeline.

I am not 100% happy with the API of `CdkDeployAction`. It currently accepts `stacks: string[]` which is the low-level API we need for the boot pipeline, but for the app pipelines, I imagine users would want to be able to pass in a `Stack` object. I wonder if we should create `CdkDeployStackAction` for that purpose.

Other remaining work:

- [ ] End-to-end test of an app with a "complex" pipeline inside it. 
- [ ] Rename the concept of bootstrapping pipeline to something nicer. I was thinking that maybe "boot pipeline" is good (just shorter).
- [ ] Update READMEs in app-delivery and pipeline-actions so they reflect that latest APIs.
- [ ] Integ test for app-delivery which instantiates the bootstrapping pipeline
- [ ] Integ test for pipeline-actions (basically define a "deployment pipeline" for a CDK app with a few stacks). We need to be creative how we can integ test this...
- [ ] Unit tests for all actions and boot pipeline.
- [ ] Go through PR comments (up here) and address them
- [ ] I think we should rename the tool from `cdk-pipelines` to `cdk-boot`.

eladb commented:
# seed

eladb commented:
🌱

skinny85 commented:
Did the rebase. Let's see the if the build passes.

EricEdens commented on packages/@aws-cdk/app-delivery/README.md:
```diff
@@ -1,154 +1,132 @@
+To use it, create a file called `cdk.pipelines.yaml` with a map where the key is
+the name of the bootstrap pipeline and the value is an object with the following options:
+
+* `source`: the GitHub repository to monitor. Must be in the form **http://github.com/ACCOUNT/REPO**.
```
Would CodeCommit work here?

ChrisLahaye replied:
Yes works for me

skinny85 commented:
Hey @eladb ,

I've been working on this recently, and I had an idea that I wanted to run by you.

For the build part of the Pipeline, I was thinking that it maybe makes sense to have a separate Action class for each language we support in the CDK. So, we would have a `JavaScriptCdkBuildAction`, `JavaCdkBuildAction`, `PythonCdkBuildAction`, etc., that would choose the correct CodeBuild image and have the correct commands to compile/lint your app, according to the way we set up your CDK project in our init templates.

Thoughts on this idea?

eladb commented:
Why do we need opinionated about build? Ideally we want people to be able to use any build system they want.

I think the build step should be fully controlled by the user, and the output artifact should be the contents of the directory produced by “cdk synth -o outdir”. This is what we call the cloud assembly. 

Bear in mind that At the moment the output produced by synth is not a self contained cloud assembly, but I’ve actually worked on this during my flight and I hope that I will be able to send a PR next week that allows you to run “cdk deploy” and point to a pre-synthesized cloud assembly.

Then, the deploy stage(s) should consume this artifact and call “cdk deploy” in an environment that we fully control and the user can not customize, for security reasons. 

Let’s touch base on this early next week? Monday?

eladb commented:
@skinny85 here's the work to decouple synth and deploy https://github.com/awslabs/aws-cdk/pull/2636

skinny85 commented:
> Why do we need opinionated about build? Ideally we want people to be able to use any build system they want.

Well, in your proposal, you have an optional 'install command' property, which defaults to `npm install`, and an optional 'build command' property, which defaults to `npm run build`. Everybody using Java, C#, Python, etc. will have to override those two properties, and also select the correct CodeBuild image to use for their build.

It's not about being opinionated, it's about convenience, and sensible defaults.

eladb commented:
Got it. I like the idea of providing higher level actions with defaults for various build environments for codebuild in general, irrespective of the cdk CI/cd

eladb commented:
This should be now unblocked.

statik commented:
This looks really cool! I'm excited to try out this feature.

eladb commented:
Superseded by #3437
